### PR TITLE
JDBC - Bugfix - Negative timestamp fix

### PR DESCRIPTION
**Pull request by Jens-H**
This PR solves the problem that negative timestamps (before epoch) with a sub-second part are not handled correctly.

I discovered this after extending the DuckDBTimestamp class for timestamp with time zone support. Therefore these changes are already included too, but so far unused.

### CODE REVIEW

**Review by hawkfish — APPROVED** [review]

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! One minor comment, otherwise looks good to me.

File: tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java [hunk 1]
```diff
@@ -64,10 +79,19 @@ public String toString() {
 	}
 
 	private static long micros2seconds(long micros) {
-		return micros / 1000_000L;
+		if ((micros % 1000_000L) >= 0) {
```
[the remark this answers is not in the document]
Comment (Jens-H) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java:82]: The else tree must be used if the sub-second part (= micros % 1000000) is negative. For this I believe the modulo is necessary to handle the edge case of 0 ns, e.g. -1_000_000 ns.

Different math may handle the edge case in the formula, but I do not see how ...

File: tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java [hunk 2]
```diff
@@ -64,10 +79,19 @@ public String toString() {
...
+			return micros / 1000_000L;
+		} else {
+			return (micros / 1000_000L) -1;
+		}
 	}
 
 	private static int nanosPartMicros(long micros) {
-		return (int) ((micros % 1000_000L) * 1000);
+		if ((micros % 1000_000L) >= 0) {
```
[the remark this answers is not in the document]
Comment (Jens-H) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java:90]: Same as above.

File: tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mytherin) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java:82]: Perhaps I'm missing something, but `micros` is a single number. The sub-second part cannot be negative by itself. Either the entire number is negative or it is positive. The modulo here will only return a number smaller than 0 if micros itself is smaller than zero, no?

[the remark this answers is not in the document]
Comment (hawkfish) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java:82]: That's apparently [not true in Java](https://www.javatpoint.com/java-modulo)

[the remark this answers is not in the document]
Comment (Jens-H) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java:82]: The problem is the edge case that micros is negative but the sub-second part not (but 0):
-1_000_000 => negative overall, but sub-second part not

[the remark this answers is not in the document]
Comment (Mytherin) [reply tools/jdbc/src/main/java/org/duckdb/DuckDBTimestamp.java:82]: I see, alright, fair enough. Thanks for the explanation!

**Pull request merged**